holidays file enhancements

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Feature Description
* Holidays file has information about prime/non-prime windows for days of the week for a site. Right now, there's no documented behavior for what happens if data for any of the days is missing or if the data is inconsistent. Also, the behavior for when the year entry is missing is not clearly defined. The general use case, as well as the theme of the prime-time feature in PBS, seems to be to assume prime time if data is insufficient, so it seems like a good idea to officially enforce this behavior. It is also desired to have an empty/commented out holidays file which doesn't need to be updated every year so that it is all prime-time by default.

#### Affected Platform(s)
* All

#### Design
* EDD: https://pbspro.atlassian.net/wiki/spaces/PD/pages/1046151225/Holidays+file+missing+inconsistent+days+will+be+treated+as+ALL+PRIME
* Forum discussion:
http://community.pbspro.org/t/holidays-file-missing-inconsistent-days-will-be-treated-as-all-prime/1440
http://community.pbspro.org/t/wrong-holidays-file-makes-scheduler-infinite-loop/1387/

#### Solution Description
* Now, after parsing holidays file, if there's data missing for a day, we enforce 'all' prime-time for it
* If an entry in holidays file sets both prime and non-prime start times to 'all'/'none', we enforce 'all' prime-time for it.
* The holidays file will now be completely commented out by default, so it'll be 24x7 prime-time

#### Testing logs/output
* [testholidays.log](https://github.com/PBSPro/pbspro/files/2850263/testholidays.log)

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [X] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [X] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented on src/scheduler/prime.c:
```diff
@@ -228,6 +228,34 @@ is_holiday(long date)
+ * @return void
+ */
+static void
+handle_missing_prime_info(void)
```
This should happen in init_config()

agrawalravi90 replied:
I wanted to do the same thing, but, if somebody adds "all all" as prime and non prime start in holidays file, then we need to be able to catch that in parse_holidays and set all prime-time, not all non-prime time. We don't have a good way of catching that if we set the bits to "all prime" inside init_config() itself as we won't be able to tell inside load_day() whether the bits are all prime because the holidays file has "all" entry for prime, or if it is what was set by init_config(). So, we could refactor the function if you really hate this, but otherwise i can't put it inside init_config().

bhroam replied:
I don't understand.  You call handle_missing_prime_info() right at the top of parse_holidays().  There should be no difference between the same code being called in init_config() or in parse_holidays()

agrawalravi90 replied:
No, I call it at the end after parsing the holidays file

bhroam commented on test/tests/functional/pbs_holidays.py:
```diff
@@ -0,0 +1,148 @@
+        jid1 = self.server.submit(j1)
+
+        # Verify that the job does not run as it is 24hrs prime-time
+        self.server.expect(JOB, {'job_state': 'Q'}, jid1)
```
Check to see that comment is set before checking for job_state: Q.  There is a race that you do your expect() before the scheduler even sees the job.

bhroam commented on src/scheduler/prime.c:
```diff
@@ -467,16 +499,30 @@ load_day(enum days d, enum prime_time pr, char *tok)
+				/*
+				 * The user specified both all prime and all non-prime for this day
+				 * all prime wins in such a situation, so just ignore the non-prime info
+				 */
```
Turn this comment into a more concise log message.  This is an error case, so we shouldn't just assume something without saying it.

agrawalravi90 replied:
Just to be sure, you want me to actually log a message to sched logs here, or just trim the comment?

bhroam replied:
The comment as a comment was fine.  I want a scheduler log message.  If we're adding a scheduler log message, I don't think we need both a comment and the log message.  For a log message, the comment is a bit wordy.

bhroam commented on src/scheduler/prime.c:
```diff
@@ -467,16 +499,30 @@ load_day(enum days d, enum prime_time pr, char *tok)
+			if (pr == NON_PRIME && conf.prime[d][PRIME].none == TRUE) {
+				/*
+				 * The user specified both none prime and none non-prime for this day
+				 * Make it all prime-time for this day as we don't know better
```
Same as above about the log message.

bhroam commented on src/scheduler/data_types.h:
```diff
@@ -1069,7 +1069,7 @@ struct config
 	time_t decay_time;			/*  time in seconds for the decay period*/
 	time_t sync_time;			/* time between syncing usage to disk */
-	struct t prime[HIGH_DAY][HIGH_PRIME];	/* prime time start and prime time end*/
+	struct t prime[HIGH_DAY][2];	/* prime time start and prime time end*/
```
Why the change from HIGH_PRIME to 2?

agrawalravi90 replied:
the other 2 rows (corresponding to ALL and NONE) were not used, so it was just unused memory.

bhroam replied:
This will probably never happen, but the idea was if we ever add more time windows than just prime and non-prime, this will just automatically update itself.  We'll add a new value to the enum and HIGH_PRIME will increase.  The amount of lost memory is pretty minor.

agrawalravi90 commented:
@bhroam thanks for your comments, I've addressed them. I also made some additional changes according to the comments on the design from the forums. Please provide feedback.

bhroam commented on test/tests/functional/pbs_holidays.py:
```diff
@@ -71,15 +71,39 @@ def check_prime(self):
+
+    def check_nonprime(self):
+        """
+        Helper function to test that it is non-prime time right now
```
If the log_filter isn't filtering DEBUG2, the scheduler will tell you what prime status it is in and when it ends in the log.  Consider using the log message.

bhroam commented on test/tests/functional/pbs_holidays.py:
```diff
@@ -146,3 +170,117 @@ def test_only_year(self):
+        """
+        Test that scheduler assumes all prime time if the year entry is
+        missing from holidays file
+        """
```
You really have tested this above.  You are testing what the scheduler will do with a completely commented out holidays file.  Above, the year was comment out, and here the commented out year is missing.  In either case the entire file is commented out.

agrawalravi90 replied:
I know, but I'd like to keep it just to be thorough.

bhroam replied:
Then make the test different.  Set non-prime such that it should be in non-prime, but since the year is missing it will be 24hr prime.

agrawalravi90 replied:
Actually I think the test already does that, it sets 24x7 non-prime time and then checks that it's 24x7 prime-time because of the missing year entry

bhroam replied:
It has 3 lines that are commented out.  I'm saying don't have them commented out.  Have them set non-prime.  This way we can easily check if 24hr prime is working.

agrawalravi90 replied:
D'oh! Those lines were never meant to commented out, sorry about that, yes, the test would be very much the same as the previous one with them commented out :) I've uncommented them now, please let me know if it looks good now.

bhroam commented on src/scheduler/pbs_holidays:
```diff
@@ -34,44 +34,25 @@
 *
-YEAR  2018
+* UNCOMMENT AND CHANGE THIS TO THE CURRENT YEAR
+*YEAR  1970
```
As I said in the forum post, use '#' to comment out the lines.  The '#'  always worked, but '*' was the correct comment character for this file to be compliant with UNICOS.

agrawalravi90 replied:
Ok, I'll make it '#' and keep the check for '*' for backward compatibility.

btw, I actually don't see your comment in the forum, could you please post it once more? Thanks.

bhroam commented on test/tests/functional/pbs_holidays.py:
```diff
@@ -146,3 +170,117 @@ def test_only_year(self):
+        # Verify that scheduler didn't log a message for out of date file
+        msg = "holidays;The holiday file is out of date; please update it."
+        self.scheduler.log_match(msg, starttime=ctime, existence=False,
+                                 max_attempts=5)
```
One of the use cases of commenting out the holiday file is so the scheduler doesn't add prime status events to the calendar.  Consider writing a test to check for that.  To do it, you need to set log_filter to 2048.  You then add a job that takes up all the resources without requesting walltime.  You then add a second job that will get calendared.  You look in the log that no policy change events happened.

agrawalravi90 commented:
@bhroam I've addressed your comments, please provide further feedback. Thanks!

bhroam commented:
Codacy is complaining about an unused variable in your test.

bhroam commented:
My only comment is the one I mentioned above for the test without a YEAR line.

agrawalravi90 commented:
@bhroam thanks for the review, I've addressed your comments. I also decided to add 2 more PTL tests to cover some corner cases. Please review the latest commit and provide feedback. Thanks.

bhroam commented on test/tests/functional/pbs_holidays.py:
```diff
@@ -334,10 +332,64 @@ def test_prime_events_calendar(self):
+        self.scheduler.holidays_set_day(today, "none", "all")
+
+        h_content = self.du.cat(filename=self.scheduler.holidays_file,
+                                sudo=True)['out']
```
Why are you catting the file?  It isn't used anywhere.

bhroam commented:
My only comment is about the test with the YEAR line.  I'm fine after we come to consensus about that test.

agrawalravi90 commented:
Thanks a lot for the thorough review Bhroam!
@arungrover could you please review this PR as well?

agrawalravi90 commented on src/scheduler/prime.c:
```diff
@@ -467,16 +501,26 @@ load_day(enum days d, enum prime_time pr, char *tok)
-		else if (!strcmp(tok, "none") || !strcmp(tok, "NONE")) {
-			conf.prime[d][pr].none = TRUE;
+			conf.prime[d][pr].none = FALSE;
+		} else if (!strcmp(tok, "none") || !strcmp(tok, "NONE")) {
```
I'm not sure if values like NoNe or NOne are supposed to be allowed. Since you left it up to me, I'd like to keep the code as is for now. Thanks for being diligent :)

mike0042 replied:
+1 for strcasecmp()
+2 for strncasecmp()

agrawalravi90 commented:
@arungrover and @bhroam thanks for the sign-offs guys, I've rebase the branch, please merge it if possible. Thanks!

bhroam commented:
Thanks @agrawalravi90.  I've merged in the PR.
